src/transfers.rs

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017-2018 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or
+// modify it under the terms of the GNU General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but
+// WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see
+// <http://www.gnu.org/licenses/>.
+
+//! [NTFS](https://github.com/CanalTP/navitia/blob/dev/documentation/ntfs/ntfs_fr.md)
+//! format management.
+
+use collection::{Collection, CollectionWithId};
+use objects::{StopPoint, Transfer};
+
+/// Checks if a transfer already exist between 2 stop points
+fn transfer_exists(transfers: &Collection<Transfer>, sp1: &StopPoint, sp2: &StopPoint) -> bool {
+    transfers
+        .values()
+        .any(|t| t.from_stop_id == sp1.id && t.to_stop_id == sp2.id)
+}
+
+/// Generates missing transfers
+///
+/// The `max_distance` argument allows you to specify the max distance
+/// in meters to compute the tranfer.
+///
+/// The `walking_speed` argument is the walking speed in meters per second.
+pub fn generates_transfers(
+    transfers: &mut Collection<Transfer>,
+    stop_points: &CollectionWithId<StopPoint>,
+    max_distance: f64,
+    walking_speed: f64,
+) {
+    for sp1 in stop_points.values() {
+        for sp2 in stop_points.values() {
+            if !transfer_exists(&transfers, &sp1, &sp2) {
+                let distance = sp1.coord.distance_to(&sp2.coord);
+                let transfer_time = (distance / walking_speed) as u32;
+                if distance <= max_distance {
+                    transfers.push(Transfer {
+                        from_stop_id: sp1.id.clone(),
+                        to_stop_id: sp2.id.clone(),
+                        min_transfer_time: Some(transfer_time),
+                        real_min_transfer_time: Some(transfer_time),
+                        equipment_id: None,
+                    });
+                }
+            }
+        }
+    }
```

Comment:
Minimizing distance computations.

```rust
let transfers_set = make_transfers_set(&transfers, &stop_points);
let sq_max_distance = max_distance * max_distance;
for (idx1, sp1) in &stop_points {
    let approx = sp1.coord.approx();
    for (_, sp2) in stop_points.values().filter(|&(idx2, _)| !transfers_set.contains((idx1, idx2))) {
        let sq_distance = approx.sq_distance_to(&sp2.coord);
        if sq_distance > sq_max_distance {
            continue;
        }
        let transfer_time = (sq_distance.sqrt() / walking_speed) as u32;
        transfers.push(Transfer {
            from_stop_id: sp1.id.clone(),
            to_stop_id: sp2.id.clone(),
            min_transfer_time: Some(transfer_time),
            real_min_transfer_time: Some(transfer_time),
            equipment_id: None,
        });
    }
}
```

Reply:
maybe just filtering by distance first before filtering with existing transfers would be enough.